src/Hosting/Hosting/src/Internal/HostingMetrics.cs

```diff
@@ -0,0 +1,174 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Metrics;
+
+namespace Microsoft.AspNetCore.Hosting;
+
+internal sealed class HostingMetrics : IDisposable
+{
+    public const string MeterName = "Microsoft.AspNetCore.Hosting";
+
+    private readonly Meter _meter;
+    private readonly UpDownCounter<long> _currentRequestsCounter;
+    private readonly Histogram<double> _requestDuration;
+
+    public HostingMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.CreateMeter(MeterName);
+
+        _currentRequestsCounter = _meter.CreateUpDownCounter<long>(
+            "current-requests",
+            description: "Current requests");
+
+        _requestDuration = _meter.CreateHistogram<double>(
+            "request-duration",
+            description: "Request duration");
+    }
+
+    public void RequestStart(string scheme, string method, HostString host)
+    {
+        // Tags must match request end.
+        var tags = CreateRequestTags(scheme, method, host);
+        _currentRequestsCounter.Add(1, tags);
+    }
+
+    public void RequestEnd(string scheme, string method, HostString host, string? routePattern, int statusCode, Exception? exception, IList<KeyValuePair<string, object?>>? customTags, long startTimestamp, long currentTimestamp)
+    {
+        // Tags must match request start.
+        var tags = CreateRequestTags(scheme, method, host);
+        _currentRequestsCounter.Add(-1, tags);
+
+        // Add information gathered during request.
+        tags.Add("status-code", GetBoxedStatusCode(statusCode));
+        if (routePattern != null)
+        {
+            tags.Add("route", routePattern);
+        }
+        // This exception is only present if there is an unhandled exception.
+        // Exceptions caught by ExceptionHandlerMiddleware and DeveloperExceptionMiddleware add exception-name to custom tags.
+        if (exception != null)
+        {
+            tags.Add("exception-name", exception.GetType().Name);
+        }
+        if (customTags != null)
+        {
+            for (var i = 0; i < customTags.Count; i++)
+            {
+                tags.Add(customTags[i]);
+            }
+        }
+
+        var duration = Stopwatch.GetElapsedTime(startTimestamp, currentTimestamp);
+        _requestDuration.Record(duration.TotalMilliseconds, tags);
+    }
+
+    public void Dispose()
+    {
+        _meter.Dispose();
+    }
+
+    public bool IsEnabled() => _currentRequestsCounter.Enabled || _requestDuration.Enabled;
+
+    private static TagList CreateRequestTags(string scheme, string method, HostString host)
```

Comment:
nit: `TagList` is kind of a big `struct` would there be any benefit to return by reference?

```csharp
private static void CreateRequestTags(string scheme, string method, HostString host, out TagList tags) {}
```

Reply:
`ref return` might offer a small benefit.

Remember that these counters are called a couple of times in the context of an HTTP request. The additional overhead is probably measured in nanoseconds.